import_obfuscation.go

```diff
@@ -0,0 +1,572 @@
+func garbleSymbolName(symName string, privImports privateImports, garbledImports map[string]string, sb *strings.Builder) string {
+	prefix, name, skipSym := splitSymbolPrefix(symName)
+	if skipSym {
+		// log.Printf("\t\t? Skipped symbol: %s", symName)
+		return symName
+	}
+
+	var namedataSym bool
+	if prefix == "type..namedata." {
+		namedataSym = true
+	}
+
+	var off int
+	for {
+		o, l := privateImportIndex(name[off:], privImports, namedataSym)
+		if o == -1 {
+			if sb.Len() != 0 {
+				sb.WriteString(name[off:])
+			}
+			break
+		}
+
+		sb.WriteString(name[off : off+o])
+		sb.WriteString(hashImport(name[off+o:off+o+l], garbledImports))
+		off += o + l
+	}
+
+	if sb.Len() == 0 {
+		// log.Printf("\t\t? Skipped symbol: %s", symName)
+		return symName
+	}
+	defer sb.Reset()
+
+	return prefix + sb.String()
+}
+
+// if symbols have one of these prefixes, skip
+// garbling
+// TODO: skip compiler generated/builtin symbols
+var skipPrefixes = []string{
+	// these symbols never contain import paths
+	"gclocals.",
+	"gclocals·",
+
+	// string names be what they be
+	"go.string.",
+
+	// skip debug symbols
+	"go.info.",
+
+	// skip entrypoint symbols
+	"main.init.",
+	"main..stmp",
+}
+
+// symbols that are related to the entrypoint
+// that cannot be garbled
+var entrypointSyms = [...]string{
+	"main.main",
+	"main..inittask",
+}
+
+// if any of these strings are found in a
+// symbol name, it should not be garbled
+var skipSubstrs = [...]string{
+	// skip test symbols
+	"_test.",
+}
+
+// prefixes of symbols that we will garble,
+// but we split the symbol name by one of
+// these prefixes so that we do not
+// accidentally garble an essential prefix
+var symPrefixes = [...]string{
+	"go.builtin.",
+	"go.itab.",
+	"go.itablink.",
+	"go.interface.",
+	"go.map.",
+	"gofile..",
+	"type..eq.",
+	"type..eqfunc.",
+	"type..hash.",
+	"type..importpath.",
+	"type..namedata.",
+	"type.",
+}
+
+// splitSymbolPrefix returns the symbol name prefix Go uses
+// to help designate the type of the symbol, and the rest of
+// the symbol name. Additionally, a bool is returned that
+// signifies whether garbling the symbol name should be skipped.
+func splitSymbolPrefix(symName string) (string, string, bool) {
+	if symName == "" {
+		return "", "", true
+	}
+
+	for _, prefix := range skipPrefixes {
+		if strings.HasPrefix(symName, prefix) {
+			return "", "", true
+		}
+	}
+
+	for _, entrySym := range entrypointSyms {
+		if symName == entrySym {
+			return "", "", true
+		}
+	}
+
+	for _, substr := range skipSubstrs {
+		if strings.Contains(symName, substr) {
+			return "", "", true
+		}
+	}
+
+	for _, prefix := range symPrefixes {
+		if strings.HasPrefix(symName, prefix) {
+			return symName[:len(prefix)], symName[len(prefix):], false
+		}
+	}
+
+	return "", symName, false
+}
+
+// privateImportIndex returns the offset and length of a private import
+// in symName. If no private imports from privImports are present in
+// symName, -1, 0 is returned.
```

Comment:
Why not return a normal error?

Reply:
It either finds a private import path in the symbol name, or it doesn't. Easier to just return -1, 0 similar to how strings.Index does. It can't really 'error' any other way